content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md

```diff
@@ -178,6 +192,273 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
 
 If you later recreate the same CustomResourceDefinition, it will start out empty.
 
+## Specifying a structural schema
+
+{{< feature-state state="beta" for_kubernetes_version="1.15" >}}
+
+CustomResources traditionally store arbitrary JSON (next to `apiVersion`, `kind` and `metadata`, which is validated by the API server implicitly). With [OpenAPI v3.0 validation](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#validation) a schema can be specified, which is validated during creation and updates, compare below for details and limits of such a schema.
+
+With `apiextensions.k8s.io/v1` the definition of a structural schema will be mandatory for CustomResourceDefinitions, while in `v1beta1` this is still optional.
+
+A structural schema is an [OpenAPI v3.0 validation schema](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#validation) which:
+
+1. specifies a non-empty type (via `type` in OpenAPI) for the root, for each specified field of an object node (via `properties` or `additionalProperties` in OpenAPI) and for each item in an array node (via `items` in OpenAPI), with the exception of:
+   * a node with `x-kubernetes-int-or-string: true`
+   * a node with `x-kubernetes-preserve-unknown-fields: true`
+2. specifies each field in an object and each item in an array which is specified within an `allOf`, `anyOf`, `oneOf` or `not` also outside of those.
```

Comment:
I'm not sure yet how I would state it more clearly, but I had to re-read this point and cross-reference it with the examples multiple times to understand the rule being described. I'm not sure whether that's a reflection of quality of the sentence or my own ignorance of the subject matter.

Is it true that property names declared in `allOf` (etc.) must also be declared in the root `properties`?

Reply:
```yaml
allOf:
- properties:
    foo: X
```

means that you have to specify `foo` at the root as well:

```yaml
properties:
  foo: Y
allOf:
- properties:
    foo: X
```

Y can differ form X, i.e. only contain the type.

If you have a better way to formulate that, let me know.

Reply:
And this is transitive for any depth and including arrays, e.g.

```yaml
allOf:
- items:
    properties:
      foo: X
```

means that you have to specify `[].foo` at the root as well:

```yaml
items:
  properties:
    foo: Y
allOf:
- items:
    properties:
      foo: X
```

Reply:
Luckily you get all of this for free with a generator like kubebuilder's controller-gen because it is just the way it work by traversing the API types.

Reply:
I also had to re-read this a couple of times (it became concise once you got the idea). Perhaps break the sentence into smaller pieces, or move the subject around?

maybe something (ugly) like
> for each field in an object and each item in an array which is specified within a logical junctor (`allOf`, `anyOf`, `oneOf` or `not`), the schema also specifies the field/item outside of those logical junctors"

or we can use these examples as well

Reply:
added the examples and took @roycaihw sentence.